### PR TITLE
Add `willReadFrequently` option when getContext("2d")

### PR DESCRIPTION
**Pull request by onsummer**
Resolved a warning on the last few versions of the browser:

image

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @onsummer!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/main/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/main/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

**Comment by ggetz** [comment]
Hi @onsummer, thanks for the PR!

We had some previous discussion around this warning on the forum in these two threads:
* https://community.cesium.com/t/canvas2d-multiple-readback-operations-using-getimagedata-are-faster-with-the-willreadfrequently-attribute-set-to-true/20703
* https://community.cesium.com/t/how-to-eliminate-cesium-warning-canvas2d-multiple-readback-operations-using-getimagedata-are-faster-with-the-willreadfrequently-attribute-set-to-true/20727

Basically, we should document performance with and without this change to ensure that it's the right one. Would you be able to post some performance tests results in this PR?

@lilleyse Do you have any additional thoughts?

**Comment by onsummer** [comment]
@ggetz Busy for working recently and do not notice those discussion, I do not know how to estimate the performance with and without this option, can u please give me some tips? Thanks.

**Comment by onsummer** [comment]
@ggetz Got a blog from cesium.com - [Entity Api Performance](https://cesium.com/blog/2018/06/21/entity-api-performance/), Can I test the performance of a large number of LabelEntity using the method described in this blog?

**Comment by ggetz** [comment]
@onsummer Yes, that should work! Getting the average FPS will also be helpful:

```
viewer.scene.debugShowFramesPerSecond = true;
```

**Comment by onsummer** [comment]
> @onsummer Yes, that should work! Getting the average FPS will also be helpful:
> 
> ```
> viewer.scene.debugShowFramesPerSecond = true;
> ```

OKay, I will test later

**Comment by onsummer** [comment]
@ggetz I use that code and reduce label's count to 1500:

```js
const viewer = new Cesium.Viewer("cesiumContainer");
viewer.scene.debugShowFramesPerSecond = true;

const randomCoordinateJitter = (degree, margin) => {
  return degree + margin * (Math.random() - 0.5) / 0.5;
};
const nCats = 1500;
for (let i = 0; i <= nCats; i++) {
  viewer.entities.add({
    position: Cesium.Cartesian3.fromDegrees(
      randomCoordinateJitter(-77, 0.5),
      randomCoordinateJitter(38.88, 0.5),
    ),
    label: {
      text: `CUTE #${i}`,
    },
    billboard: {
      height: 24,
      width: 36,
      image: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRhFpaxDXCS5O9hx90F3ufJI2VnC_wW0lPnrr6BIb18P4V5JXxBCg',
    },
    ellipse: {
      semiMajorAxis: 100,
      semiMinorAxis: 80,
    }
  });
}
```

Regardless of whether `willReadFrequently` option is used, when scene rendering stabilizes, I observe that frame rate is roughly around 39~41 fps on my machine. (R7 4800U with 16GB RAM, not using discrete gpu)

Here are the testing profile file:

[Profile-WithOrWithoutFrequentlyOption.zip](https://github.com/CesiumGS/cesium/files/9872367/Profile-WithFrequentlyOption.zip)

with option

without option

Combining the above statistical data, I thought about it, `willReadFrequently` may be helpful for those 2d canvas read pixel operations frequently, but the 2 performance tests seem that maybe there is no (or almost no) 2d canvas read-pixels operation? I'm not sure if my assumption is correct.

**Comment by ctr1key** [comment]
Adding some helpful information to consider should it help this effort.
There is a [link](https://github.com/whatwg/html/issues/5614#issuecomment-866944763) to a document written by the Chrome team about the history of needing willReadFrequently and its performance impact.
From what I gather:

1.  It should not be used on canvases where you never expect to read data from the canvas with getImageData()/toDataUrl()/toBlob(). Using {willReadFrequently: true} will disable hardware acceleration for that canvas.

2. Without using willReadFrequently, it will use GPU rendering for draw commands before the first call to getImageData() and CPU rendering for subsequent draw commands. 
The impact to performance of getImageData() when using {willReadFrequently: true} vs {willReadFrequently: undefined} is only seen during the first call on a canvas.
After the first call to getImageData(), the performance of getImageData() is the same whether you use {willReadFrequently: true} or {willReadFrequently: undefined}.
Using {willReadFrequently: false} will incurred the performance hit of getImageData() for every call.
3. The performance gain of using getImageData() without hardware acceleration {willReadFrequently: true} is roughly equivalent to performance loss of doing approximately 30000 draw commands with CPU rendering instead of GPU rendering.
Using {willReadFrequently: false} will prevent the fallback to CPU rendering in a case of heaving drawing and infrequent calls to getImageData() can improve performance.
5. It's not expected to see a difference in FPS if you are making less than 4 getImageData() calls in a single frame when the canvas is hardware accelerated.  At one call per frame getImageData(), it may be worse to not use {willReadFrequently: false} as you give up other benefits of hardware acceleration.
6. You will not see a performance difference regarding the use of willReadFrequently if hardware acceleration is already disabled globally for all canvases.

**Comment by ctr1key** [comment]
As for you example of 1500 static label, you may not see much performance difference because I believe only 16 total canvases are used for your entities.
1 canvas is used per character in a label. 6 canvases for 'CUTE #' and 10 canvases for the digits. I don't believe there is a canvas for the billboard.
I believe labels call getImageData() twice so 32 calls (16 slow, then 16 quick after it falls back to CPU rendering) with {willReadFrequently: undefined} vs 32 quick calls with {willReadFrequently: true} is not much difference.
It will likely be briefly noticeable if you have some way to dynamically change traits of all labels like font family, style, or weight. A change to any of those will force a new canvas and calls to getImageData().

Aside from canvases for labels and getimagepixels, I'm not sure if the other places you used {willReadFrequently: true} will even use getImageData(). It will have an adverse effect by disabling hardware acceleration with no benefit if they don't need to use getImageData().
